src/OpenDiffix.Core/NodeUtils.fs

```diff
@@ -113,11 +113,15 @@ let inline collectAggregates node =
 
   node |> collect exprAggregators
 
-/// Visits all aggregate expressions in a node.
+/// Visits all aggregate expressions being immediate children of `node`.
 let inline visitAggregates f node =
   let rec exprVisitor f expr =
     match expr with
     | FunctionExpr (AggregateFunction (_fn, _opts), _args) as aggregateExpression -> f aggregateExpression
     | other -> other |> visit (exprVisitor f)
 
   node |> visit (exprVisitor f)
+
+let rec visitIncludingSubQueries visitor (query: SelectQuery) =
```

Comment:
Visit always includes subqueries, maybe `visitIncludingSelf`?

Reply:
`visitRecursive`?.. But I'm not convinced either...

Reply:
I would move this to a private function where it's required and keep the interface of this module with only basic blocks.

Reply:
Actually, that's what I started with. I'll revert :+1:

Reply:
It takes one line to generalize validators for subqueries, something like:

```fsharp
let rec validateQuery query =
  // ... validate this query

  query |> visit validateQuery // descend to subqueries
```

Or the opposite if you want bottom-up traversal.
Maybe drop it altogether?